### PR TITLE
Add HeroMode widget

### PR DESCRIPTION
**Pull request by najeira**
## Description

When there are multiple Navigators like CupertinoPageScaffold, I want to disable heroes in the page that is not displayed.

HeroMode widget provides a means to control whether heroes is enabled or disabled.

### Error

current Flutter, An error occurs in the following step:

* Nested Navigators, such as CupertinoPageScaffold
* Inner CupertinoTabViews has the same name Heroes
* Show both inner CupertinoTabViews to init
* Push route that has the same name Hero to root Navigator

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

const String kHeroTag = "hero tag";

void main() => runApp(CupertinoApp(
      title: 'Hero with CupertinoTabScaffold',
      home: HomePage(),
    ));

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return CupertinoTabScaffold(
      tabBar: CupertinoTabBar(
        items: <BottomNavigationBarItem>[
          BottomNavigationBarItem(icon: Icon(Icons.home)),
          BottomNavigationBarItem(icon: Icon(Icons.favorite)),
        ],
      ),
      tabBuilder: (BuildContext context, int index) {
        return CupertinoTabView(
          builder: (BuildContext context) => FirstPage(),
        );
      },
    );
  }
}

class FirstPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      navigationBar: CupertinoNavigationBar(),
      child: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Hero(
              tag: kHeroTag,
              child: Container(
                width: 100,
                height: 100,
                color: Colors.red,
              ),
            ),
            RaisedButton(
              onPressed: () {
                Navigator.of(context, rootNavigator: false).push(
                  CupertinoPageRoute(
                    builder: (BuildContext context) => NextPage(),
                  ),
                );
              },
              child: Text("Next in tab"),
            ),
            RaisedButton(
              onPressed: () {
                Navigator.of(context, rootNavigator: true).push(
                  CupertinoPageRoute(
                    builder: (BuildContext context) => NextPage(),
                  ),
                );
              },
              child: Text("Next out tab"),
            ),
          ],
        ),
      ),
    );
  }
}

class NextPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      navigationBar: CupertinoNavigationBar(),
      child: Center(
        child: Hero(
          tag: kHeroTag,
          child: Container(
            width: 100,
            height: 100,
            color: Colors.red,
          ),
        ),
      ),
    );
  }
}
```

## Related Issues

https://github.com/flutter/flutter/pull/44890
https://github.com/flutter/flutter/issues/28042

## Tests

I added the following tests:

* test/widgets/heroes_test.dart
  * Can push/pop on outer Navigator if nested Navigators contains same Heroes
  * Heroes in enabled HeroMode do transition
  * Heroes in disabled HeroMode do not transition

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

This is the addition of a new widget, will not break existing apps.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by justinmc — COMMENTED** [review]
Sorry this took so long to get a review!  Here are a few minor changes that I saw.  I checked out the branch and confirmed that the tests properly test the added functionality.

We should get an approval of the overall strategy from @goderbauer since you guys were talking about this approach in the previous PR.

**Comment by justinmc** [comment]
@najeira Are you still available to help finish up this PR?

**Comment by najeira** [comment]
Yes! Thanks for the review. Sorry for noticing late.

I will update.

**Comment by justinmc** [comment]
Cool, no problem!  Thanks for taking a look at my review.

**Comment by najeira** [comment]
Some tests have failed. Should I rebase?

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/heroes.dart]: This could inherit from ProxyWidget instead?

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/heroes.dart]: This should document what it means that Heros are enabled/disabled. e.g. if it is disabled all Hero widgets in this subtree will not be considered for Hero animations or something like that.

**Review by goderbauer — COMMENTED** [review]
Approach seems fine.

**Comment by chunhtai** [comment]
I cannot reproduce the original error, and https://github.com/flutter/flutter/issues/28042 seems to be fixed.
What is this pr aim to fix?

**Comment by najeira** [comment]
Explanation of reproduction procedure was not enough.

* launch app
* show heart tab
* push "Next out tab"

error:

```
════════ Exception caught by scheduler library ═════════════════════════════════════════════════════
The following assertion was thrown during a scheduler callback:
There are multiple heroes that share the same tag within a subtree.

Within each subtree for which heroes are to be animated (i.e. a PageRoute subtree), each Hero must have a unique non-null tag.
In this case, multiple heroes had the following tag: hero tag

Here is the subtree for one of the offending heroes: Hero
  tag: hero tag
  dependencies: [_ModalScopeStatus]
  state: _HeroState#b2991
When the exception was thrown, this was the stack: 
#0      Hero._allHeroesFor.inviteHero.<anonymous closure> (package:flutter/src/widgets/heroes.dart:266:11)
#1      Hero._allHeroesFor.inviteHero (package:flutter/src/widgets/heroes.dart:277:8)
#2      Hero._allHeroesFor.visitor (package:flutter/src/widgets/heroes.dart:305:23)
#3      MultiChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:5929:16)
#4      Hero._allHeroesFor.visitor (package:flutter/src/widgets/heroes.dart:309:15)
...
════════════════════════════════════════════════════════════════════════════════════════════════════
```

**Comment by goderbauer** [comment]
@chunhtai can you reproduce this with the additional steps?

**Comment by justinmc** [comment]
I got a crash when I followed those steps on the iOS simulator, and I confirmed that this PR fixes it.

**Comment by chunhtai** [comment]
I see, yes i can reproduce it now. Can you open a new issue about this? It is fixing a different problem from  https://github.com/flutter/flutter/issues/28042

File: packages/flutter/lib/src/cupertino/tab_scaffold.dart
Comment (chunhtai) [inline packages/flutter/lib/src/cupertino/tab_scaffold.dart]: This seems like a bandage fix.

I am not sure if this make sense, but should we apply HeroMode in Offstage directly? For example if Offstage.offstage = true, the it will wrap its child with HeroMode. 

cc @goderbauer @justinmc

Comment (goderbauer) [reply packages/flutter/lib/src/cupertino/tab_scaffold.dart]: No, it should be composable. Just like Offstage also doesn't wrap TickerMode and some other widgets.

Comment (chunhtai) [reply packages/flutter/lib/src/cupertino/tab_scaffold.dart]: sounds good

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/heroes.dart]: nit: this have to be a full sentence.
Maybe combine both sentences.
`If this property is false, the [Hero]es in this subtree will not animate on route changes; Otherwise, they will animate as usual`

File: packages/flutter/test/widgets/heroes_test.dart
Comment (chunhtai) [inline packages/flutter/test/widgets/heroes_test.dart]: I might miss something, but is this really related to nested navigators? I thought the problem is the CupertinoTabScaffold builds a stack where some of the children are offstage and the heroes in those children conflicting with each other?

File: packages/flutter/test/widgets/heroes_test.dart
Comment (chunhtai) [inline packages/flutter/test/widgets/heroes_test.dart]: @justinmc i am a bit confused about the architecture in CupertinoTabScaffold. Why does CupertinoTabView need a navigator? It seems like the tab is controlled by a stack in CupertinoTabScaffold.

Comment (chunhtai) [reply packages/flutter/test/widgets/heroes_test.dart]: never mind, I think i figured out

**Comment by najeira** [comment]
Thanks for the review!

I'm not good at English. I'm not sure how to fix any documentation issues.

I will check and respond to issues regarding the implementation.

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (justinmc) [reply packages/flutter/lib/src/widgets/heroes.dart]: @najeira Try this.  Make sure to break any lines that exceed the 80 character width limit.


Enables or disables [Hero]es in the widget subtree.

When [enabled] is false, all Hero widgets in this subtree will not be considered for Hero animations.

When [enabled] is true (the default), Hero widgets may be involved in a Hero animation, like normal.

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (najeira) [reply packages/flutter/lib/src/widgets/heroes.dart]: ProxyWidget is abstract, the code addition will increase. SingleChildRenderObjectWidget is good for this, right?

Comment (najeira) [reply packages/flutter/lib/src/widgets/heroes.dart]: Is it better to inherit ProxyWidget and return SingleChildRenderObjectElement?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart]: No, SingleChildRenderObjectWidget is expected to have a renderObject, The HeroMode itself does not layout or paint in the render layer. It is only adding information to the widget tree. The ProxyWidget returns ComponentElement might make most sense in this case.

Note: there is a ProxyElement, but it has specific logic built in to notify child when its configuration changes. We do not need this functionality, so i think ProxyElement is not suitable in this use case

**Comment by justinmc** [comment]
@najeira How's it going with these last few comments?  Let me know if anything isn't clear about the documentation and I can help out by writing the English.

**Comment by najeira** [comment]
updated comments

**Review by justinmc — APPROVED** [review]
LGTM but @chunhtai and/or @goderbauer should take a final look before we merge.  There are one or two comments that I'm not sure have been addressed.

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart:900]: This seems not be addressed yet?

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart:917]: same here

File: packages/flutter/test/widgets/heroes_test.dart
Comment (chunhtai) [reply packages/flutter/test/widgets/heroes_test.dart:1928]: same here

**Review by chunhtai — COMMENTED** [review]
Some of the comments have not been addressed yet

**Comment by najeira** [comment]
@chunhtai Try reloading.

Then you can see this:

スクリーンショット 2020-04-28 10 29 32

**Comment by chunhtai** [comment]
> @chunhtai Try reloading.
> 
> Then you can see this:
> 
> スクリーンショット 2020-04-28 10 29 32

We were discussing about using ProxyWidget instead of StatelessWidget

**Comment by najeira** [comment]
Sorry for late. I will update it!

**Comment by najeira** [comment]
I updated HeroMode to extends InheritedWidget.

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/heroes.dart:910]: It should not be an inherited widget. The inherited widget will cause unnecessary rebuild when its flag changes.

I think It should be ProxyWidget

Comment (najeira) [reply packages/flutter/lib/src/widgets/heroes.dart:910]: Only a few widgets below inherit ProxyWidget.

- ParentDataWidget
- InheritedWidget

I am still not convinced if ProxyWidget is appropriate for HeroMode. I think that it seems too fundamental and unsuitable for inheritance.

If we use ProxyWidget, we also need to implement Element. It seems overkill to me for what we are trying to do.

If updateShouldNotify always returns false, then it can avoid unnecessary rebuild?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart:910]: no, inherited widget adds additional overhead to to the element tree(all the inherited widget will be cached on each element in the tree, see Element._inheritedWidgets), and it is not something we should use if we don't needs its inherited functionality.

Although proxy widget is not used as often as other other type of widget, it seems to be the best choice for this case without adding additional overhead

Comment (najeira) [reply packages/flutter/lib/src/widgets/heroes.dart:910]: I understood that InheritedWidget has overhead.

How about StatelessWidget? I think StatelessWidget is good for this.

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart]: stateless widget will introduce an additional call to the build method, if you use proxy widget you can directly return the child. Although it may not be a big deal, but why not perfect it?

Comment (najeira) [reply packages/flutter/lib/src/widgets/heroes.dart]: If we use ProxyWidget, we need to implement HeroModeElement. It is big deal than introducing build method.

The following widgets inherit StatelessWidget and do `Widget build(BuildContext context) => child;`, not inherit ProxyWidget.

- NotificationListener
- KeyedSubtree
- PageStorage
- PreferredSize

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart]: yes you will need to implement HeroModeElement which you can extend from the ComponentElement, in that class you just need to be sure to implement build and update method. You can take a look at StatelessElement

Comment (najeira) [reply packages/flutter/lib/src/widgets/heroes.dart]: Is using ProxyWidget faster than using StatelessWidget? Are there any other benefits?

OK, I will try!

Comment (najeira) [reply packages/flutter/lib/src/widgets/heroes.dart]: If we implement HeroModeElement in the same way as StatelessElement, we can use StatelessWidget instead of ProxyWidget. I do not understand why to use ProxyWidget. What is the difference between HeroModeElement and StatelessElement?

**Comment by najeira** [comment]
Could anybody show me an example using ProxyWidget and ProxyElement?

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart]: Not much difference except the build method of HeroModeElement does not do a `widget.build(context)`, instead it just return widget.child. This save us a function call.

**Comment by chunhtai** [comment]
> Could anybody show me an example using ProxyWidget and ProxyElement?

The ProxyElement is for inheritedElement that it implement extra logic

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (najeira) [reply packages/flutter/lib/src/widgets/heroes.dart]: StatelessElement uses Element._dirty, HeroModeElement can not use the variable because it class is in the outside. And it is not good to implement HeroModeElement in widgets/framework.dart.

**Comment by najeira** [comment]
Could anybody show me an example using ProxyWidget and ComponentElement?

**Comment by najeira** [comment]
Is `Widget build(BuildContext context) => child;` a heavy process to avoid?

**Comment by najeira** [comment]
In my understanding, implementing HeroModeElement skips the call HeroMode.build.

extends StatelessWidget:

- StatelessElement.build
- HeroMode.build
- returned child widget is used

extends ProxyWidget and impelement HeroModeElement:

- HeroModeElement.build
- returned child widget is used

Is this all? Please let me know if there are any other processes that affect performance.

**Comment by chunhtai** [comment]
> In my understanding, implementing HeroModeElement skips the call HeroMode.build.
> 
> extends StatelessWidget:
> 
> * StatelessElement.build
> * HeroMode.build
> * returned child widget is used
> 
> extends ProxyWidget and impelement HeroModeElement:
> 
> * HeroModeElement.build
> * returned child widget is used
> 
> Is this all? Please let me know if there are any other processes that affect performance.

yes that is all as far as I know



>Is Widget build(BuildContext context) => child; a heavy process to avoid?

It is not a heavy process, but we can improve it by implementing the HeroModeElement which is not too much work.

**Comment by najeira** [comment]
`Widget build(BuildContext context) => child;` is very very small task,  it is hard to find even in performance measurement. It does not allocate any widgets, just returns variable. I think that implementing the HeroModeElement is too much work to avoid its micro task. 

If we should optimize it, it is better to do it in another PR together with the following classes:

- NotificationListener
- KeyedSubtree
- PageStorage
- PreferredSize

These widgets inherit StatelessWidget and do `Widget build(BuildContext context) => child;`, not inherit ProxyWidget. They are the same as this case.



Additionally: what do you think of this comment?

> StatelessElement uses Element._dirty, HeroModeElement can not use the variable because it class is in the outside. And it is not good to implement HeroModeElement in widgets/framework.dart.

----

I want you to comment on which is preferable.

CC: @justinmc @goderbauer

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart]: I see, then it seems like the only way to use the proxy widget will be overriding the StatelessElement and override the build method.

I agree this may make the code looks weird and probably not worth the effort to the save one build method. yeah, let's use stateless widget, what you have looks good.

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart:927]: this seems not addressed yet?

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/heroes.dart:935]: this can be an arrow function

File: packages/flutter/test/widgets/heroes_test.dart
Comment (chunhtai) [reply packages/flutter/test/widgets/heroes_test.dart:2027]: This is not addressed yet?

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/heroes.dart:908]:
```suggestion
/// When [enabled] is true (the default), [Hero] widgets may be involved in
```

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/heroes.dart:909]:
```suggestion
/// hero animations, as usual.
```

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/heroes.dart:905]:
```suggestion
/// When [enabled] is false, all [Hero] widgets in this subtree will not be
```

File: packages/flutter/lib/src/widgets/heroes.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/heroes.dart:906]:
```suggestion
/// involved in hero animations.
```

**Review by chunhtai — APPROVED** [review]
LGTM, with some nit pick

File: packages/flutter/test/widgets/heroes_test.dart
Comment (najeira) [reply packages/flutter/test/widgets/heroes_test.dart:2027]: Heroes are animated by screen transition. This is related to Navigator. The problem is caused by multiple nested Navigators.

**Comment by najeira** [comment]
diff: https://github.com/najeira/flutter/compare/70815c70ecdec2088ec047c6c46da70afcb527b9...a76023bc172ef77e251b14404dadb83a2947d5e7

**Comment by chunhtai** [comment]
@najeira it seems like this branch is outdated, can you try to rebase off latest master and see if you can fix the cirrus?

**Comment by najeira** [comment]
I rebased this branch.

**Comment by chunhtai** [comment]
@najeira 
I am not sure what is going on with the presubmit, but we have some infra change in the last couple of days. Can you rebase off the latest master again and see if it makes cirrus happy?

**Comment by najeira** [comment]
I am very grateful for the long discussions and for helping me with my poor English. Thank you all.

**Comment by justinmc** [comment]
Just saw that this was merged, thank you @najeira!

**Pull request merged**
